jormungandr-watchdog/src/service/intercom.rs

```diff
@@ -0,0 +1,268 @@
+                        );
+                        connection.sent_counter.fetch_add(1, Ordering::SeqCst);
+                        r
+                    }
+                    _ => {
+                        tracing::debug!("service not connected");
+                        Err(msg)
+                    }
+                };
+
+            if retry.is_err() && retry_attempted {
+                tracing::error!("cannot connect to service");
+                return Err(WatchdogError::CannotConnectToService {
+                    service_identifier: T::SERVICE_IDENTIFIER,
+                    retry_attempted,
+                });
+            }
+
+            if retry.is_err() {
+                retry_attempted = true;
+                tracing::debug!("retrying to connect to service in 100ms");
+                tokio::time::delay_for(std::time::Duration::from_millis(100)).await;
+                self.connect().in_current_span().await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn disconnect(&mut self) {
+        let span = tracing::span!(tracing::Level::DEBUG, "Intercom::disconnect");
+        let _enter = span.enter();
+        tracing::trace!("disconnect from the service");
+        self.state = IntercomState::Disconnected;
+    }
+
+    async fn connect(&mut self) -> Result<(), WatchdogError> {
+        let span = tracing::span!(tracing::Level::DEBUG, "Intercom::connect");
+        let _enter = span.enter();
+
+        // make sure we are disconnected
+        self.disconnect();
+
+        let (reply, receiver) = oneshot::channel();
+
+        let command = ControlCommand::Intercom {
+            service_identifier: T::SERVICE_IDENTIFIER,
+            reply: Reply(reply),
+        };
+        tracing::trace!("querying connection to service from the watchdog");
+        self.watchdog_query.send(command).await;
+
+        match receiver.await {
+            Ok(Ok(intercom_sender)) => {
+                tracing::trace!("watchdog replied with established connection");
+                let tid = intercom_sender.type_id();
+                match intercom_sender.downcast_ref::<IntercomSender<T::Intercom>>() {
+                    Some(intercom_sender_ref) => {
+                        self.state = IntercomState::Connected {
+                            connection: intercom_sender_ref.clone(),
+                        };
+                        Ok(())
+                    }
+                    None => unreachable!(
+                        "cannot downcast the intercom object to {}, {:?}",
+                        std::any::type_name::<T::Intercom>(),
+                        tid,
+                    ),
+                }
+            }
+            Ok(Err(err)) => {
+                tracing::error!(error = %err, "cannot connect to the service");
+                Err(err)
+            }
+            Err(err) => {
+                // we assume the server will always reply one way or another
+                unreachable!(
```

Comment:
If this can occur in real life, better use `panic!`?

Reply:
it should not and if it does, it is a bug